### PR TITLE
Update Java Lab websocket protocol to use JSON

### PR DESCRIPTION
**Pull request by jmkulwik**
Javabuilder is updating its websocket protocol to use formatted json. This will allow us to do the following:
* translate our messages from javabuilder
* route signals from javabuilder to the appropriate section of javalab
* define a clear communication protocol between javabuilder and javalab

Some details about the internationalization approach: https://docs.google.com/document/d/1gBWh1b9rH1DRpheKph1I1KjT-melx01yB0xXs6g2GTU/edit#

Pairs with this PR from the javabuilder repo: https://github.com/code-dot-org/javabuilder/pull/32

## Links

- spec: [Neighborhood engineering plan](https://docs.google.com/document/d/1SqSLogXsOKJD8QexbZAbxCpUGRdNP9Yj7zW8tN6oADI/edit#)
- jira ticket: [CSA-287](https://codedotorg.atlassian.net/browse/CSA-287)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/javalab/javabuilderExceptionHandler.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  const type = exceptionDetails.value;
+  const detail = exceptionDetails.detail;
+  let args = {};
+  if (detail.connectionId) {
+    args.connectionId = detail.connectionId;
+  }
+  if (detail.cause) {
+    args.cause = detail.cause;
+  }
+  callback(window.locales.javalab_locale[type](args));
```
Comment (jmkulwik) [inline apps/src/javalab/javabuilderExceptionHandler.js]: ~Here, we map directly from [keys set in javabuilder](https://github.com/code-dot-org/javabuilder/pull/32/files#diff-588560c58930937a4accd4ce0b91566398194bff4efdff36953ed6acf997f14c) to i18n messages.~

File: apps/src/javalab/JavabuilderConnection.js [hunk 2]
```diff
@@ -37,32 +41,55 @@ export default class JavabuilderConnection {
...
-      this.onOutputMessage('Compiling...');
-    };
+  onOpen() {
+    this.onOutputMessage('Compiling...');
+  }
 
-    this.socket.onmessage = event => {
-      this.onOutputMessage(event.data);
-    };
+  onMessage(event) {
```
Comment (jmkulwik) [inline apps/src/javalab/JavabuilderConnection.js:56]: This is the only method that is functionally different in Javabuilder Connection

File: apps/src/javalab/JavabuilderConnection.js [hunk 3]
```diff
@@ -37,32 +41,55 @@ export default class JavabuilderConnection {
...
+        this.onOutputMessage(data.value);
+        break;
+      case WebSocketMessageType.NEIGHBORHOOD:
+        Neighborhood.handleSignal(data);
+        break;
+      case WebSocketMessageType.EXCEPTION:
+        handleException(data, this.onOutputMessage);
+        break;
+      case WebSocketMessageType.DEBUBG:
+        this.onOutputMessage('--- Localhost debugging message ---');
```
Comment (molly-moen) [inline apps/src/javalab/JavabuilderConnection.js]: will students ever see a debug message or is this only for us?

File: apps/test/util/javalab/locale-do-not-import.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/**
+ * DO NOT IMPORT THIS DIRECTLY. Instead do:
+ *   ```
+ *   import msg from '@cdo/javalab/locale'.
+ *   ```
+ * This allows the webpack config to determine how locales should be loaded,
+ * which is important for making locale setup work seamlessly in tests.
+ */
+// make sure Blockly is loaded
```
Comment (molly-moen) [inline apps/test/util/javalab/locale-do-not-import.js]: remove this comment?

**Review by molly-moen — APPROVED** [review]
Exciting! Just a couple clarifying comments, nothing blocking from my side

File: apps/test/util/javalab/locale-do-not-import.js (hunk 4 above)
Comment (jmkulwik) [reply apps/test/util/javalab/locale-do-not-import.js]: Haha oops. Thanks!

File: apps/src/javalab/JavabuilderConnection.js (hunk 3 above)
Comment (jmkulwik) [reply apps/src/javalab/JavabuilderConnection.js]: This is only for us. I'll add a check to make sure we never accidentally print this.

**Pull request merged**